modules/gdscript/language_server/gdscript_workspace.cpp

```diff
@@ -0,0 +1,504 @@
+	ClassDB::bind_method(D_METHOD("parse_local_script", "p_path"), &GDScriptWorkspace::parse_local_script);
+	ClassDB::bind_method(D_METHOD("get_file_path", "p_uri"), &GDScriptWorkspace::get_file_path);
+	ClassDB::bind_method(D_METHOD("get_file_uri", "p_path"), &GDScriptWorkspace::get_file_uri);
+	ClassDB::bind_method(D_METHOD("publish_diagnostics", "p_path"), &GDScriptWorkspace::publish_diagnostics);
+	ClassDB::bind_method(D_METHOD("generate_script_api", "p_path"), &GDScriptWorkspace::generate_script_api);
+}
+
+void GDScriptWorkspace::remove_cache_parser(const String &p_path) {
+	Map<String, ExtendGDScriptParser *>::Element *parser = parse_results.find(p_path);
+	Map<String, ExtendGDScriptParser *>::Element *script = scripts.find(p_path);
+	if (parser && script) {
+		if (script->get() && script->get() == script->get()) {
+			memdelete(script->get());
+		} else {
+			memdelete(script->get());
+			memdelete(parser->get());
+		}
+		parse_results.erase(p_path);
+		scripts.erase(p_path);
+	} else if (parser) {
+		memdelete(parser->get());
+		parse_results.erase(p_path);
+	} else if (script) {
+		memdelete(script->get());
+		scripts.erase(p_path);
+	}
+}
+
+const lsp::DocumentSymbol *GDScriptWorkspace::get_native_symbol(const String &p_class, const String &p_member) const {
+
+	StringName class_name = p_class;
+	StringName empty;
+
+	while (class_name != empty) {
+		if (const Map<StringName, lsp::DocumentSymbol>::Element *E = native_symbols.find(class_name)) {
+			const lsp::DocumentSymbol &class_symbol = E->value();
+
+			if (p_member.empty()) {
+				return &class_symbol;
+			} else {
+				for (int i = 0; i < class_symbol.children.size(); i++) {
+					const lsp::DocumentSymbol &symbol = class_symbol.children[i];
+					if (symbol.name == p_member) {
+						return &symbol;
+					}
+				}
+			}
+		}
+		class_name = ClassDB::get_parent_class(class_name);
+	}
+
+	return NULL;
+}
+
+const lsp::DocumentSymbol *GDScriptWorkspace::get_script_symbol(const String &p_path) const {
+	const Map<String, ExtendGDScriptParser *>::Element *S = scripts.find(p_path);
+	if (S) {
+		return &(S->get()->get_symbols());
+	}
+	return NULL;
+}
+
+void GDScriptWorkspace::reload_all_workspace_scripts() {
+	List<String> pathes;
+	list_script_files("res://", pathes);
+	for (List<String>::Element *E = pathes.front(); E; E = E->next()) {
+		const String &path = E->get();
+		Error err;
+		String content = FileAccess::get_file_as_string(path, &err);
+		ERR_CONTINUE(err != OK);
+		err = parse_script(path, content);
+
+		if (err != OK) {
+			Map<String, ExtendGDScriptParser *>::Element *S = parse_results.find(path);
+			String err_msg = "Failed parse script " + path;
+			if (S) {
+				err_msg += "\n" + S->get()->get_error();
+			}
+			ERR_EXPLAIN(err_msg);
+			ERR_CONTINUE(err != OK);
+		}
+	}
+}
+
+void GDScriptWorkspace::list_script_files(const String &p_root_dir, List<String> &r_files) {
+	Error err;
+	DirAccessRef dir = DirAccess::open(p_root_dir, &err);
```

Comment:
Traversing the file system might be really slow on large project. I think it would be better to get the `EditorFileSystem` cache.

Reply:
This is running in a seperated thread so I think it would be ok for performance.